torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -261,49 +318,128 @@ void PropagateShapeOnNode(Node * node, bool insert_expands) {
     default:
       break; // fall-through
   }
-  if (node->matches("aten::cat(Tensor[] tensors, int dim) -> Tensor", /*with_const=*/attr::dim)) {
-    auto list_node = node->namedInput(attr::tensors)->node();
-    JIT_ASSERT(list_node->kind() == prim::ListConstruct);
-    auto tensors = list_node->inputs();
-    if (tensors.size() > 0) {
-      auto input_types = fmap(tensors, [](Value *v) { return v->type()->cast<TensorType>(); });
-      if (std::all_of(input_types.begin(), input_types.end(),
-          [](const TensorTypePtr& tp) { return tp != nullptr; })) {
-        std::vector<int64_t> sizes = input_types[0]->sizes();
-        const int64_t dim = wrapDim(node->get<int64_t>(attr::dim).value(), sizes);
-        const int64_t ndim = sizes.size();
-
-        if (dim < 0 || dim >= ndim)
-          goto cat_fail;
-
-        sizes[dim] = 0;
-        for (auto & tp : input_types) {
-          auto & tp_sizes = tp->sizes();
-          if (sizes.size() != tp_sizes.size())
-            goto cat_fail;
-          for (int64_t i = 0; i < ndim; ++i) {
-            if (sizes[i] != tp_sizes[i] && i != dim) {
-              goto cat_fail;
-            }
-          }
-          sizes[dim] += tp_sizes[dim];
+  if (node->matches("aten::cat(Tensor[] tensors, int dim) -> Tensor")) {
+    return PropagateCatShape(node);
+  }
+
+  if (auto maybe_complete_types = gatherTensorTypes<CompleteTensorType>(node)) {
+    if (PropagateCompleteShapeOnNode(node, insert_expands, std::move(*maybe_complete_types))) {
+      return;
+    }
+  }
+
+  if (auto maybe_tensor_types = gatherTensorTypes<TensorType>(node)) {
+    if (PropagateTensorShapeOnNode(node, insert_expands, std::move(*maybe_tensor_types))) {
+      return;
+    }
+  }
+
+  if (canPropagateShapeByRunningIt(node))
```

Comment:
Can we just make PropagageShapeOnNodeByRunning it return a boolean like the others?

Reply:
Sure.